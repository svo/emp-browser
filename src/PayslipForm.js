import React, { Component } from 'react';
import { FormGroup, ControlLabel, FormControl } from 'react-bootstrap';
import PropTypes from 'prop-types';

class PayslipForm extends Component {

  static propTypes = {
    year: PropTypes.number.isRequired,
    month: PropTypes.number.isRequired
  }

  constructor(props) {
    super(props);

    var date = new Date(this.props.year, this.props.month);
    var locale = "en-au";
    var month = date.toLocaleString(locale, { month: "long" });

    this.state = {
      month: month.toUpperCase(),
      first_name: "",
      last_name: ""
    };

    this.handleFirstNameChange = this.handleFirstNameChange.bind(this);
    this.handleLastNameChange = this.handleLastNameChange.bind(this);
  }

  validFirstName() {
    const length = this.state.first_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  validLastName() {
    const length = this.state.last_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  handleFirstNameChange(e) {
    this.setState({ first_name: e.target.value });
  }

  handleLastNameChange(e) {
    this.setState({ last_name: e.target.value });
  }

  render() {
    return (
      <form>
        <FormGroup controlId="first_name" validationState={this.validFirstName()}>
          <ControlLabel>First Name</ControlLabel>
          <FormControl type="text"
            placeholder="Sean"
            value={this.state.first_name}
            onChange={this.handleFirstNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="last_name" validationState={this.validLastName()}>
          <ControlLabel>Last Name</ControlLabel>
          <FormControl type="text"
            placeholder="Van Osselaer"
            value={this.state.last_name}
            onChange={this.handleLastNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="annual_salary">
          <ControlLabel>Annual Salary</ControlLabel>
          <FormControl type="text" placeholder="175000" />
        </FormGroup>
        <FormGroup controlId="year">
          <ControlLabel>Year</ControlLabel>
          <FormControl type="text"
                       defaultValue={this.props.year}
                       placeholder={this.props.year} />
        </FormGroup>
        <FormGroup controlId="month">
          <ControlLabel>Month</ControlLabel>
          <FormControl componentClass="select" placeholder="..."
              defaultValue={this.state.month}>
            <option value="JANUARY">January</option>
            <option value="FEBRUARY">February</option>
            <option value="MARCH">March</option>
            <option value="APRIL">April</option>
            <option value="MAY">May</option>
            <option value="JUNE">June</option>
            <option value="JULY">July</option>
            <option value="AUGUST">August</option>
            <option value="SEPTEMBER">September</option>
            <option value="OCTOBER">October</option>
            <option value="NOVEMBER">November</option>
            <option value="DECEMBER">December</option>
          </FormControl>
        </FormGroup>
        <FormGroup controlId="super_rate">
          <ControlLabel>Super Rate</ControlLabel>
          <FormControl type="text" placeholder="..."
            defaultValue="9"/>
        </FormGroup>
      </form>
    );
  }
}

export default PayslipForm;
