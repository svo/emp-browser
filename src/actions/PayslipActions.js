import alt from '../alt';

class PayslipActions {
  create(details) {
    return "/";
  }
}

export default alt.createActions(PayslipActions);
