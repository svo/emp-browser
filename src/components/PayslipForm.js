import React, { Component } from 'react';
import { FormGroup, ControlLabel, FormControl, Button, ButtonGroup } from 'react-bootstrap';
import PropTypes from 'prop-types';
import format from 'date-format';
import PayslipActions from '../actions/PayslipActions';
import PayslipStore from '../stores/PayslipStore';

class PayslipForm extends Component {

  static propTypes = {
    now: PropTypes.object.isRequired
  }

  constructor(props) {
    super(props);

    this.state = {
      first_name: "",
      last_name: "",
      annual_salary: 0,
      month: this.props.now,
      super_rate: 9
    };

    this.handleFirstNameChange = this.handleFirstNameChange.bind(this);
    this.handleLastNameChange = this.handleLastNameChange.bind(this);
    this.handleAnnualSalaryChange = this.handleAnnualSalaryChange.bind(this);
    this.handleMonthChange = this.handleMonthChange.bind(this);
    this.handleSuperRateChange = this.handleSuperRateChange.bind(this);
    this.handleCreateClick = this.handleCreateClick.bind(this);
    this.onChange = this.onChange.bind(this);
  }

  getInitialState() {
    return PayslipStore.getState();
  }

  componentDidMount() {
    PayslipStore.listen(this.onChange);
  }

  componentWillUnmount() {
    PayslipStore.unlisten(this.onChange);
  }

  onChange(state) {
    this.setState({location: state.payslip.location});
  }

  validFirstName() {
    const length = this.state.first_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  validLastName() {
    const length = this.state.last_name.length;
    return (length > 0) ? 'success' : 'error';
  }

  validAnnualSalary() {
    var value = parseFloat(this.state.annual_salary);

    if (typeof value === 'number' &&
      isFinite(value) &&
      Math.floor(value) === value &&
      value > 0) {

      return 'success';
    } else {
      return 'error';
    }
  }

  validSuperRate() {
    var value = parseFloat(this.state.super_rate);

    if (typeof value === 'number' &&
      isFinite(value) &&
      value >= 0 &&
      value <= 50) {

      return 'success';
    } else {
      return 'error';
    }
  }

  valid() {
    return this.validSuperRate() === 'success' &&
      this.validAnnualSalary() === 'success' &&
      this.validFirstName() === 'success' &&
      this.validLastName() === 'success';
  }

  handleFirstNameChange(e) {
    this.setState({ first_name: e.target.value });
  }

  handleLastNameChange(e) {
    this.setState({ last_name: e.target.value });
  }

  handleAnnualSalaryChange(e) {
    this.setState({ annual_salary: e.target.value });
  }

  handleMonthChange(e) {
    this.setState({ month: new Date(e.target.value) });
  }

  handleSuperRateChange(e) {
    this.setState({ super_rate: e.target.value });
  }

  handleCreateClick() {
    PayslipActions.create(this.state);
  }

  render() {
    return (
      <form>
        <FormGroup controlId="first_name" validationState={this.validFirstName()}>
          <ControlLabel>First Name</ControlLabel>
          <FormControl type="text"
            placeholder="Sean"
            value={this.state.first_name}
            onChange={this.handleFirstNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="last_name" validationState={this.validLastName()}>
          <ControlLabel>Last Name</ControlLabel>
          <FormControl type="text"
            placeholder="Van Osselaer"
            value={this.state.last_name}
            onChange={this.handleLastNameChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="annual_salary" validationState={this.validAnnualSalary()}>
          <ControlLabel>Annual Salary</ControlLabel>
          <FormControl type="number"
            placeholder="175000"
            value={this.state.annual_salary}
            onChange={this.handleAnnualSalaryChange} />
          <FormControl.Feedback />
        </FormGroup>
        <FormGroup controlId="year">
          <ControlLabel>Year</ControlLabel>
          <FormControl type="month"
            value={format('yyyy-MM', this.state.month)}
            onChange={this.handleMonthChange} />
        </FormGroup>
        <FormGroup controlId="super_rate" validationState={this.validSuperRate()}>
          <ControlLabel>Super Rate</ControlLabel>
          <FormControl type="number"
            placeholder="9"
            value={this.state.super_rate}
            onChange={this.handleSuperRateChange} />
          <FormControl.Feedback />
        </FormGroup>
        <ButtonGroup>
          <Button bsStyle="primary" onClick={this.handleCreateClick}>
            Create
          </Button>
          <Button bsStyle="success" href={this.state.location}
            disabled={this.state.location === undefined}>
            Download
          </Button>
        </ButtonGroup>
      </form>
    );
  }
}

export default PayslipForm;
