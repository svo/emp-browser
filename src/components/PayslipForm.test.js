import React from 'react';
import ReactDOM from 'react-dom';
import PayslipForm from './PayslipForm';
import PayslipActions from '../actions/PayslipActions';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <PayslipForm now={new Date()} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

describe("properties", function() {
  it('should have correct now property', () => {
    var now = new Date();
    const div = document.createElement('div');

    var payslipForm = ReactDOM.render(<PayslipForm now={now} />, div);

    expect(payslipForm.props.now).toEqual(now);
  });
});

describe("state", function() {
  it('should have correct first name default', () => {
    const div = document.createElement('div');

    var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

    expect(payslipForm.state.first_name).toEqual("");
  });

  it('should have correct last name default', () => {
    const div = document.createElement('div');

    var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

    expect(payslipForm.state.last_name).toEqual("");
  });

  it('should have correct annual salary default', () => {
    const div = document.createElement('div');

    var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

    expect(payslipForm.state.annual_salary).toEqual(0);
  });

  it('should have correct month state', () => {
    const div = document.createElement('div');
    var now = new Date();

    var payslipForm = ReactDOM.render(<PayslipForm now={now} />, div);

    expect(payslipForm.state.month).toEqual(now);
  });
});

describe("validation", function() {
  it("should be valid", () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      payslipForm.setState({super_rate: 9,
        annual_salary: 175000,
        first_name: "Sean",
        last_name: "Van Osselaer"});

      expect(payslipForm.valid()).toEqual(true);
  });

  it("should be invalid if super rate invalid", () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      const validSuper = jest.fn();
      validSuper.mockReturnValue('error');
      payslipForm.validSuperRate = validSuper;

      expect(payslipForm.valid()).toEqual(false);
  });

  it("should be invalid if annual salary invalid", () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      payslipForm.setState({super_rate: 9});
      const validAnnualSalary = jest.fn();
      validAnnualSalary.mockReturnValue('error');
      payslipForm.validAnnualSalary = validAnnualSalary;

      expect(payslipForm.valid()).toEqual(false);
  });

  it("should be invalid if first name invalid", () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      payslipForm.setState({super_rate: 9,
        annual_salary: 175000});
      const validFirstName = jest.fn();
      validFirstName.mockReturnValue('error');
      payslipForm.validFirstName = validFirstName;

      expect(payslipForm.valid()).toEqual(false);
  });

  it("should be invalid if last name invalid", () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      payslipForm.setState({super_rate: 9,
        annual_salary: 175000,
        first_name: "Sean"});
      const validLastName = jest.fn();
      validLastName.mockReturnValue('error');
      payslipForm.validLastName = validLastName;

      expect(payslipForm.valid()).toEqual(false);
  });

  describe("of first name", function() {
    it('should report invalid first name', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      expect(payslipForm.validFirstName()).toEqual('error')
    });

    it('should report valid first name', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({first_name: "Sean"});

      expect(payslipForm.validFirstName()).toEqual('success')
    });
  });

  describe("of last name", function() {
    it('should report invalid last name', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      expect(payslipForm.validLastName()).toEqual('error')
    });

    it('should report valid last name', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({last_name: "Van Osselaer"});

      expect(payslipForm.validLastName()).toEqual('success')
    });
  });

  describe("of annual salary", function() {
    it('should report invalid annual salary if not provided', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);

      expect(payslipForm.validAnnualSalary()).toEqual('error')
    });

    it('should report invalid annual salary for decimal values', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({annual_salary: "1.01"});

      expect(payslipForm.validAnnualSalary()).toEqual('error')
    });

    it('should report invalid annual salary for negative', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({annual_salary: "-1"});

      expect(payslipForm.validAnnualSalary()).toEqual('error')
    });

    it('should report valid annual salary', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({annual_salary: "1"});

      expect(payslipForm.validAnnualSalary()).toEqual('success')
    });
  });

  describe("of super rate", function() {
    it('should report invalid super rate if not provided', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({super_rate: 51});

      expect(payslipForm.validSuperRate()).toEqual('error')
    });

    it('should report invalid super rate for negative', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({super_rate: "-1"});

      expect(payslipForm.validSuperRate()).toEqual('error')
    });

    it('should report valid super rate for 0', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({super_rate: "0"});

      expect(payslipForm.validSuperRate()).toEqual('success')
    });

    it('should report valid super rate for 50', () => {
      const div = document.createElement('div');

      var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />, div);
      payslipForm.setState({super_rate: "10"});

      expect(payslipForm.validSuperRate()).toEqual('success')
    });
  });
});

describe("on create", function() {
  it('should call create action', () => {
    const mockAction = jest.fn();
    const state = {annual_salary: 175000,
      first_name: "Sean",
      last_name: "Van Osselaer",
      month: new Date(),
      super_rate: 9};
    var payslipForm = ReactDOM.render(<PayslipForm now={new Date()} />,
      document.createElement('div'));
    PayslipActions.create = mockAction;
    payslipForm.setState(state);

    payslipForm.handleCreateClick();

    expect(mockAction).toBeCalledWith(state);
  });
});
