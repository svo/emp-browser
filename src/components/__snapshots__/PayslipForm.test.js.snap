// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form>
  <div
    className="form-group has-feedback has-error"
  >
    <label
      className="control-label"
      htmlFor="first_name"
    >
      First Name
    </label>
    <input
      className="form-control"
      id="first_name"
      onChange={[Function]}
      placeholder="Sean"
      type="text"
      value=""
    />
    <span
      className="form-control-feedback glyphicon glyphicon-remove"
    />
  </div>
  <div
    className="form-group has-feedback has-error"
  >
    <label
      className="control-label"
      htmlFor="last_name"
    >
      Last Name
    </label>
    <input
      className="form-control"
      id="last_name"
      onChange={[Function]}
      placeholder="Van Osselaer"
      type="text"
      value=""
    />
    <span
      className="form-control-feedback glyphicon glyphicon-remove"
    />
  </div>
  <div
    className="form-group has-feedback has-error"
  >
    <label
      className="control-label"
      htmlFor="annual_salary"
    >
      Annual Salary
    </label>
    <input
      className="form-control"
      id="annual_salary"
      onChange={[Function]}
      placeholder="175000"
      type="number"
      value={0}
    />
    <span
      className="form-control-feedback glyphicon glyphicon-remove"
    />
  </div>
  <div
    className="form-group"
  >
    <label
      className="control-label"
      htmlFor="year"
    >
      Year
    </label>
    <input
      className="form-control"
      id="year"
      onChange={[Function]}
      type="month"
      value="2017-09"
    />
  </div>
  <div
    className="form-group has-feedback has-success"
  >
    <label
      className="control-label"
      htmlFor="super_rate"
    >
      Super Rate
    </label>
    <input
      className="form-control"
      id="super_rate"
      onChange={[Function]}
      placeholder="9"
      type="number"
      value={9}
    />
    <span
      className="form-control-feedback glyphicon glyphicon-ok"
    />
  </div>
  <button
    className="btn btn-primary"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Create
  </button>
</form>
`;
